src/metric.h

```diff
@@ -0,0 +1,177 @@
+  /// \return Returns the shared memory handle of the metric.
+  bi::managed_external_buffer::handle_t ShmHandle();
+
+  /// Get the request kind of the metric.
+  /// \return Returns the request kind of the metric.
+  const MetricRequestKind& RequestKind();
+
+  /// Get the value of the metric.
+  /// \return Returns the value of the metric.
+  double Value();
+
+  /// Save Custom Metric object to shared memory.
+  /// \param shm_pool Shared memory pool to save the custom metric object.
+  void SaveToSharedMemory(std::unique_ptr<SharedMemoryManager>& shm_pool);
+
+  /// Create a Custom Metric object from shared memory.
+  /// \param shm_pool Shared memory pool
+  /// \param handle Shared memory handle of the custom metric.
+  /// \return Returns the custom metrics in the specified request_handle
+  /// location.
+  static std::unique_ptr<Metric> LoadFromSharedMemory(
+      std::unique_ptr<SharedMemoryManager>& shm_pool,
+      bi::managed_external_buffer::handle_t handle);
+
+  /// Get the address of the TRITONSERVER_Metric object.
+  /// \return Returns the address of the TRITONSERVER_Metric object.
+  void* MetricAddress();
+
+  /// Send the request to the parent processto delete the Metric object.
+  void Clear();
+
+#ifdef TRITON_PB_STUB
+  /// Send the request to the parent process to increment the metric by the
+  /// specified value.
+  /// \param value The value to increment the metric by.
+  void SendIncrementRequest(const double& value);
+
+  /// Send the request to the parent process to set the metric to the specified
+  /// value.
+  /// \param value The value to set the metric to.
+  void SendSetValueRequest(const double& value);
+
+  /// Send the request to the parent process to get the value of the metric.
+  /// \return Returns the value of the metric.
+  double SendGetValueRequest();
+#else
+  // Initialize the TRITONSERVER_Metric object.
+  /// \return Returns the address of the TRITONSERVER_Metric object.
+  void* InitializeTritonMetric();
+
+  /// Parse the labels string into a vector of TRITONSERVER_Parameter.
+  /// \param labels_params The vector of TRITONSERVER_Parameter to store the
+  /// parsed labels.
+  /// \param labels The labels string to parse.
+  void ParseLabels(
+      std::vector<const TRITONSERVER_Parameter*>& labels_params,
+      const std::string& labels);
+
+  /// Handle the metric operation.
+  /// \param metrics_message_ptr The pointer to the CustomMetricsMessage object.
+  void HandleMetricOperation(CustomMetricsMessage** metrics_message_ptr);
+
+  /// Use Triton C API to increment the value of the metric by the given value.
+  /// \param value The value to increment the metric by.
+  void Increment(const double& value);
+
+  /// Use Triton C API to set the value of the metric to the given value.
+  /// \param value The value to set the metric to.
+  void SetValue(const double& value);
+
+  /// Use Triton C API to get the value of the metric.
+  void UpdateValue();
```

Comment:
I think this is also updating the internal value, and must be called by other operations to keep local value state correct?

Reply:
Yes, that was the purpose. However, I think we don't need to update the value after each operation as we don't directly retrieve the value from the `Metric` object. Instead, we get the value directly from the result of Triton C API call. I changed this function name to `GetValue()`, and the `value_` variable now is called `operation_value_` indicating that it's for the operation calls.